sched/sched/sched_releasetcb.c

```diff
@@ -161,6 +164,23 @@ int nxsched_release_tcb(FAR struct tcb_s *tcb, uint8_t ttype)
 
       group_leave(tcb);
 
+      /* Kernel thread and group still reference by pthread */
+
+#ifndef CONFIG_DISABLE_PTHREAD
+      if (ttype != TCB_FLAG_TTYPE_PTHREAD)
```

Comment:
```suggestion
      if (ttype == TCB_FLAG_TTYPE_PTHREAD && (tcb->flags & TCB_FLAG_FREE_TCB))
        {
          kmm_free(tcb);
        }
```
and revert the change of group_release

Reply:
group_leave will be called twice, once before nxtask_terminate() and the other time before free(tcb):


```
#0  group_leave (tcb=0x419e60 <nxtask_signalparent+119>) at group/group_leave.c:224
#1  0x0000000000419eb9 in nxtask_exithook (tcb=0x7ffff3db5330, status=0) at task/task_exithook.c:457
#2  0x0000000000452733 in nx_pthread_exit (exit_value=0x0) at pthread/pthread_exit.c:104
#3  0x000000000041f18e in pthread_exit (exit_value=0x0) at pthread/pthread_exit.c:71
#4  0x0000000000461ac2 in hello_main (argc=1, argv=0x7ffff3db5950) at hello_main.c:66
#5  0x000000000041f61c in nxtask_startup (entrypt=0x4619cb <hello_main>, argc=1, argv=0x7ffff3db5950) at sched/task_startup.c:70
#6  0x0000000000417138 in nxtask_start () at task/task_start.c:112
#7  0x000000000042579f in pre_start () at sim/sim_initialstate.c:52
#8  0x0000000000000000 in ?? ()
```

```
#0  group_leave (tcb=0x425a00 <up_release_stack+71>) at group/group_leave.c:224
#1  0x000000000041929f in nxsched_release_tcb (tcb=0x7ffff3dc61a0, ttype=1 '\001') at sched/sched_releasetcb.c:165
#2  0x000000000041a0a9 in nxtask_terminate (pid=5) at task/task_terminate.c:121
#3  0x0000000000453a9e in nxtask_exit () at task/task_exit.c:167
#4  0x0000000000425b01 in up_exit (status=0) at sim/sim_exit.c:63
#5  0x000000000045273d in nx_pthread_exit (exit_value=0x16) at pthread/pthread_exit.c:106
#6  0x000000000041f18e in pthread_exit (exit_value=0x16) at pthread/pthread_exit.c:71
#7  0x00000000004590e4 in pthread_startup (entry=0x461968 <iperf_report_task>, arg=0x0) at pthread/pthread_create.c:59
#8  0x00000000004b405c in pthread_start () at pthread/pthread_create.c:139
#9  0x000000000042579f in pre_start () at sim/sim_initialstate.c:52
#10 0x0000000000000000 in ?? ()
```

I added a task thread check in the first group_release(), because if tcb is released before nxtask_terminate(), then the subsequent flow will trigger use-after-free, and within nxtask_terminate(), both task and pthread are possible Called, so we needs check whether the tcb is not referenced by other threads.

Reply:
could we simplify the exit process?

Reply:
Done

Reply:
I kept the original implement, group_leave cannot be optimized in nxtask_exithook()
```

nx_pthread_exit
|
|-->nxtask_exithook
|   |
|   |-->group_leave
|   |   |
|   |    -->group_release /* Group resource who depends on task status which must be freed in this stage */
|   |       |
|   |        -->group_drop  /* 1. do not free the resource if thread type is task, tcb will be still referenced on up_exit */
|   |                       /* 2. if thread type is pthread, which means nobody reference on this group, drop it */
|   |
|    -->nxsig_cleanup /* Tcb reference */
|
 -->up_exit
    |
     -->nxtask_exit
        |
        |-->nxsched_add_blocked  /* TSTATE_TASK_INACTIVE */
        |
         -->nxtask_terminate
            |
             -->nxsched_release_tcb
                |
                |-->group_leave
                |   |
                |    -->group_release
                |       |
                |        -->group_drop  /* Only pthread could release the task group(tcb) */
                |
                 -->free(tcb) /* 1. If task exit before pthread, check the reference(tg_members/tg_nwaiters) and skip tcb free, which will be freed from child thread who use this group */
                              /* 2. If thread type is task, and no child thread reference on this group, free the tcb */
```